package netmsg;

message AppRequest {
  // Uniquely identifies one of our users
  required int64 phone_id = 1;

  // This is a guarantee a message of a certain type will exist.
  // At time of writing, no situations will send concurrent messagse
  enum MessageType {
    tStatusUpdate=1;
    tCreateRoom=2;
    tCreatePost=3;
    tRegistration=4;
    tLogin=5;
    tLogout=6;
  }
  required MessageType msg_type = 2;

  // Implementation done
  message RegisterMessage {
    // Already have phone_id
    required string secured_password = 1;
    required string username = 2;
    optional string email = 3;
  }
  optional RegisterMessage reg_msg = 3;

  message LoginMessage {
    // Priority over phone login
    optional string email = 1;
    // Server-side only compare hashes
    required string secured_password = 2;
  }
  optional LoginMessage login_msg = 4;

  // Message definitions for status updating
  message StatusUpdate {
    required string body = 1;
    // Repeated 0 or more times
    repeated bytes photos = 2;
  }
  optional StatusUpdate status_updates = 5;

  enum RoomType {
    Event=1; // Activity, group
    Profile=2; // User profile, contains wall
    Page=3; // Non-affiliation. Can remove this
  }

  message CreateEvent {
    required int64 host_id = 1;
    required string title = 2;
    required string location = 3;
    required int64 time = 4; // POSIX time
    repeated int64 invited_users = 5;
  }
  optional CreateEvent create_event = 6;

  // Separating messages for event accept,rejc,invt in anticipation of future diversification.
  // If there is contention for 1-16 ids, we can combine these message types.
  message EventAccept {
    required int64 user_id = 1;
    required bytes event_uuid1 = 2;
  }
  optional EventAccept accept_event = 7;

  message EventReject {
    required int64 user_id = 1;
    required bytes event_uuid1 = 2;
  }
  optional EventReject reject_event = 8;

  message EventInvite {
    required int64 user_id = 1;
    required bytes event_uuid1 = 2;
  }
  optional EventInvite invite_event = 9;

  message PollInvited {
    required int64 user_id = 1;
    optional int64 offset = 2;
    optional int64 amount = 3;
  }
  optional PollInvited poll_invited = 10;

  message PollAccepted {
    required int64 user_id = 1;
    optional int64 offset = 2;
    optional int64 amount = 3;
  }
  optional PollAccepted poll_accepted = 11;

  // Rooms represent anything with a "wall". User profiles, events, etc.
  message CreateRoom {
    required string name = 1;
    required RoomType type = 2;
    optional bytes uuid_suggestion = 3;
  }
  optional CreateRoom create_room_msg = 13;

  message RoomPostMessage {
    required bytes room_uuid = 1;
    required string body = 2;
    repeated bytes photos = 3;
  }
  optional RoomPostMessage room_post_msg = 14;

  optional bool logout_request = 15;
}

message AppReply {
  // fail -> error, error reason in message
  required bool success = 1;

  message CreateRoomReply {
    optional bytes success_uuid = 1;
  }
  optional CreateRoomReply create_reply=2;

  message CreateEventUUID {
    required bytes event_uuid1 = 1;
  }

}
